### PR TITLE
PVF: Instantiate wasm in pre-checking

### PR DESCRIPTION
**Pull request by mrcnski**
# PULL REQUEST

## Overview

Do an additional instantiation check to catch more issues.

## Relevant issues

Closes https://github.com/paritytech/polkadot/issues/7218

### CODE REVIEW

File: roadmap/implementers-guide/src/pvf-prechecking.md [hunk 1]
```diff
@@ -21,7 +21,17 @@ As a result of this issue we need a fairly hard guarantee that the PVFs of regis
...
 ## Solution
 
-The problem is solved by having a pre-checking process which is run when a new validation code is included in the chain. A new PVF can be added in two cases:
+The problem is solved by having a pre-checking process.
+
+### Pre-checking
+
+Pre-checking mostly consists of attempting to prepare (compile) the PVF WASM blob. We use more strict limits (e.g. timeouts) here compared to regular preparation for execution. This way errors during preparation later are likely unrelated to the PVF itself, as it already passed pre-checking. We can treat such errors as local node issues.
+
+We also have an additional step where we attempt to instantiate the WASM runtime without running it. This is unrelated to preparation so we don't time it, but it does help us catch more issues.
```
Comment (mrcnski) [inline roadmap/implementers-guide/src/pvf-prechecking.md:30]: - Can we now treat runtime construction errors as internal errors? I didn't do that because the PVF can still be un-instantiable on certain machines/configurations while making it through the pre-check vote, making attacks against those environments possible.

File: node/core/pvf/src/pvf.rs
Comment (mrcnski) [inline node/core/pvf/src/pvf.rs:1]: Prepare-specific code so I moved it to the `prepare` module. I also needed to reference a type that is defined there.

File: roadmap/implementers-guide/src/pvf-prechecking.md (hunk 1 above)
Comment (eskimor) [reply roadmap/implementers-guide/src/pvf-prechecking.md:30]: Technically, I think we have to have nodes abstain from voting in that case, anything else would be unfair: They voted against the PVF in pre-checking, but later we slash them because they fail validating a candidate? Does not sound fair.

On the other hand if it passed pre-checking, we know there is a super majority being able to validate the candidate, so it should be "fine" if some nodes don't vote. 

I do think we have to time the instantiation - it could be possible to make this take a very long time? 

At the same time, I would assume in the honest case for it to be very quick, so it should be ok to make this part of the preparation timeout in pre-checking.

Comment (mrcnski) [reply roadmap/implementers-guide/src/pvf-prechecking.md:30]: > I do think we have to time the instantiation - it could be possible to make this take a very long time?
>
> At the same time, I would assume in the honest case for it to be very quick, so it should be ok to make this part of > the preparation timeout in pre-checking.

We do have the long host-side timeout, but yeah that's not ideal. Lumping it in with the preparation timeout makes it harder to reason about as we are no longer comparing the same operations in prechecking vs. preparation.

Best would be to have a separate timeout for instantiation, as well as for instantiation in execution where it would be more lenient. This way we actually make sure that if it passes in prechecking it should pass in execution. And having it in a separate thread here as well as in execution means we can do more fine-grained sandboxing later. Or am I overcomplicating it? I can also just do your suggestion for now and file an issue for later.

Comment (eskimor) [reply roadmap/implementers-guide/src/pvf-prechecking.md:30]: Given that our pre-checking timeout is already way lower than the preparation timeout, it seems fine to include the instantiation - assuming we make sure the timeout is such that for honest PVFs it should not trigger. So indeed, I think we can keep it simple. It does not have to be comparable - it has to be "lower enough". Given that execution timeout is in the seconds, which includes instantiation I can hardly imagine how this simplification could be causing issues. It should be clearly documented though.

File: node/core/pvf/worker/src/executor_intf.rs [hunk 2]
```diff
@@ -227,6 +228,22 @@ impl Executor {
 		}
 		.map_err(|err| format!("execute error: {:?}", err))
 	}
+
+	/// Constructs the runtime for the given PVF.
+	///
+	/// # Safety
+	///
+	/// See `execute` for safety considerations.
```
Comment (slumber) [inline node/core/pvf/worker/src/executor_intf.rs:236]: nit
```
[`execute`]
```
to create a doc-link

File: roadmap/implementers-guide/src/pvf-prechecking.md (hunk 1 above)
Comment (mrcnski) [reply roadmap/implementers-guide/src/pvf-prechecking.md:30]: Hmm, we need the compiled artifact on-disk to run instantiation, and by that point the timeout has been stopped. We need to implement https://github.com/paritytech/substrate/issues/13861. Then we could check that the artifact is instantiable before we actually write it to disk.

Comment (burdges) [reply roadmap/implementers-guide/src/pvf-prechecking.md:30]: We'd still publish VRFs but then no-show I think.  In principle a vote against the PVF could mean "no-show early" but afaik there is no reason forthat extra code complexity, probably you just announce your VRF and then abandon it.

Comment (eskimor) [reply roadmap/implementers-guide/src/pvf-prechecking.md:30]: > Hmm, we need the compiled artifact on-disk to run instantiation, and by that point the timeout has been stopped. We need to implement [paritytech/substrate#13861](https://github.com/paritytech/substrate/issues/13861). Then we could check that the artifact is instantiable before we actually write it to disk.

Let's just make sure there is some timeout. The rest is: "What is easiest to implement" - it should not matter too much.

Comment (mrcnski) [reply roadmap/implementers-guide/src/pvf-prechecking.md:30]: @eskimor Done. Once we have sandboxing I think this will need its own thread. For now it should be good.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2875401

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2875716

File: node/core/pvf/execute-worker/src/executor_intf.rs
Comment (mrcnski) [inline node/core/pvf/execute-worker/src/executor_intf.rs:1]: Moved this to `pvf/common` since `Executor` was now used in prechecking for runtime construction!

File: node/core/pvf/common/src/executor_intf.rs [hunk 3]
```diff
@@ -112,3 +119,248 @@ pub fn params_to_wasmtime_semantics(par: &ExecutorParams) -> Result<Semantics, S
...
+	/// Executes the given PVF in the form of a compiled artifact and returns the result of execution
+	/// upon success.
+	///
+	/// # Safety
+	///
+	/// The caller must ensure that the compiled artifact passed here was:
+	///   1) produced by [`prepare`],
+	///   2) written to the disk as a file,
+	///   3) was not modified,
+	///   4) will not be modified while any runtime using this artifact is alive, or is being
```
Comment (eskimor) [inline node/core/pvf/common/src/executor_intf.rs:145]: First, I love this explicit listing of requirements. Second, I think this is obsolete, as the artifact is now passed in directly, instead of a file path. Or am I missing something?

File: node/core/pvf/common/src/executor_intf.rs [hunk 4]
```diff
@@ -112,3 +119,248 @@ pub fn params_to_wasmtime_semantics(par: &ExecutorParams) -> Result<Semantics, S
...
+type HostFunctions = (
+	sp_io::misc::HostFunctions,
+	sp_io::crypto::HostFunctions,
+	sp_io::hashing::HostFunctions,
+	sp_io::allocator::HostFunctions,
+	sp_io::logging::HostFunctions,
+	sp_io::trie::HostFunctions,
+);
+
+/// The validation externalities that will panic on any storage related access.
```
Comment (eskimor) [inline node/core/pvf/common/src/executor_intf.rs:199]: Not clear here, why we would want that.

File: node/core/pvf/common/src/executor_intf.rs [hunk 5]
```diff
@@ -112,3 +119,248 @@ pub fn params_to_wasmtime_semantics(par: &ExecutorParams) -> Result<Semantics, S
 	sem.deterministic_stack_limit = Some(stack_limit);
 	Ok(sem)
 }
+
+#[derive(Clone)]
+pub struct Executor {
```
Comment (eskimor) [inline node/core/pvf/common/src/executor_intf.rs:124]: Some docs would be nice. What is an `Executor`? Why do we need it? ..

File: node/core/pvf/execute-worker/src/lib.rs [hunk 6]
```diff
@@ -43,6 +41,42 @@ use std::{
...
+//    will abort and that will abort the process as well.
+//
+// Typically on Linux the main thread gets the stack size specified by the `ulimit` and
+// typically it's configured to 8 MiB. Rust's spawned threads are 2 MiB. OTOH, the
+// NATIVE_STACK_MAX is set to 256 MiB. Not nearly enough.
+//
+// Hence we need to increase it. The simplest way to fix that is to spawn a thread with the desired
+// stack limit.
+//
+// The reasoning why we pick this particular size is:
```
Comment (eskimor) [inline node/core/pvf/execute-worker/src/lib.rs:74]: Nice docs!

File: node/core/pvf/prepare-worker/src/lib.rs [hunk 7]
```diff
@@ -151,11 +155,23 @@ pub fn worker_entrypoint(socket_path: &str, node_version: Option<&str>) {
 			let prepare_thread = thread::spawn_worker_thread(
 				"prepare thread",
 				move || {
-					let result = prepare_artifact(pvf, cpu_time_start);
+					#[allow(unused_mut)]
```
Comment (eskimor) [inline node/core/pvf/prepare-worker/src/lib.rs:158]: Why is this necessary? (allow(unused_mut))?

**Review by eskimor — COMMENTED** [review]
Nice work @mrcnski  !

File: node/core/pvf/common/src/executor_intf.rs (hunk 3 above)
Comment (mrcnski) [reply node/core/pvf/common/src/executor_intf.rs:145]: Good catch, forgot to update that! To be clear though I didn't write these requirements. :P

File: node/core/pvf/common/src/executor_intf.rs (hunk 4 above)
Comment (mrcnski) [reply node/core/pvf/common/src/executor_intf.rs:199]: This was already there, I just moved some stuff around in this PR. But, it is a good question. I always assumed it was because validation shouldn't ever need to call these functions, though maybe there is a stronger reason. We should confirm the reasoning and document it.

File: node/core/pvf/prepare-worker/src/lib.rs (hunk 7 above)
Comment (mrcnski) [reply node/core/pvf/prepare-worker/src/lib.rs:158]: Without it, the following line on Linux results in a warning about it being unused:

```rs
#[cfg(target_os = "linux")]
let mut result = result.map(|(artifact, elapsed)| (artifact, elapsed, get_max_rss_thread()));
```

File: node/core/pvf/common/src/executor_intf.rs (hunk 5 above)
Comment (mrcnski) [reply node/core/pvf/common/src/executor_intf.rs:124]: Actually it doesn't seem that useful now that the rayon threadpool has been removed. It used to look [like this](https://github.com/paritytech/polkadot/blob//node/core/pvf/worker/src/executor_intf.rs#L146). Without that, these methods can be standalone functions that take a `Config`. I won't change it here in case it could be useful again later, but I can write a ticket.

File: node/core/pvf/execute-worker/src/lib.rs (hunk 6 above)
Comment (mrcnski) [reply node/core/pvf/execute-worker/src/lib.rs:74]: I can't take the credit. 'Twas @pepyakin. But yeah, they are inspiring. :)

**Review by eskimor — APPROVED** [review]

File: node/core/pvf/common/src/executor_intf.rs (hunk 5 above)
Comment (mrcnski) [reply node/core/pvf/common/src/executor_intf.rs:126]: https://github.com/paritytech/polkadot/issues/7313

File: node/core/pvf/common/src/executor_intf.rs (hunk 4 above)
Comment (mrcnski) [reply node/core/pvf/common/src/executor_intf.rs]: I was able to find https://github.com/paritytech/polkadot/pull/3005 with some justification, which I added as docs here. :)

(Aside: I also randomly found https://github.com/paritytech/polkadot/pull/2710 which is amazing, especially the linked hackmd.)

Comment (mrcnski) [reply node/core/pvf/common/src/executor_intf.rs]: This is off-topic but I feel like we should make the historical PRs more discoverable. Especially now that I borked the blame history with all my refactors. 🙈 Maybe a `HISTORY.md` would be nice, for context and for linking out to important PRs. And why is [this project](https://github.com/paritytech/polkadot/projects/10#card-64533754) closed I wonder? 🤔

**Review by slumber — APPROVED** [review]

**Comment by mrcnski** [comment]
bot merge
